src/core/ComboBox/ComboBox.tsx

```diff
@@ -0,0 +1,282 @@
+  const getOptionId = React.useCallback(
+    (index: number) =>
+      options[index].id ??
+      `${id}-option${options.findIndex(
+        ({ value }) => value === options[index].value,
+      )}`,
+    [options, id],
+  );
+
+  const inputRef = React.useRef<HTMLInputElement>(null);
+  const menuRef = React.useRef<HTMLUListElement>(null);
+  const [isOpen, setIsOpen] = React.useState(false);
+
+  // Set min-width of menu to be same as input
+  const [minWidth, setMinWidth] = React.useState(0);
+  React.useEffect(() => {
+    if (inputRef.current) {
+      setMinWidth(inputRef.current.offsetWidth);
+    }
+  }, [isOpen]);
+
+  const [filteredOptions, setFilteredOptions] = React.useState(options);
+  const [focusedIndex, setFocusedIndex] = React.useState<number>(-1);
+
+  // Maintain internal selected value state synced with `value` prop
+  const [selectedValue, setSelectedValue] = React.useState<T | undefined>();
+  React.useEffect(() => {
+    setSelectedValue(value);
+  }, [value]);
+
+  // Fire onChange callback every time selected value changes
+  React.useEffect(() => {
+    onChange?.(selectedValue ?? null);
+  }, [selectedValue, onChange]);
+
+  // Controlled input value
+  const [inputValue, setInputValue] = React.useState<string>('');
+  const onInput = React.useCallback(
+    (event: React.ChangeEvent<HTMLInputElement>) => {
+      setInputValue(event.target.value);
+      inputProps?.onChange?.(event);
+    },
+    [inputProps],
+  );
+
+  // Update input value when option is selected
+  React.useEffect(() => {
+    const selectedOption = options.find(({ value }) => value === selectedValue);
+    selectedOption && setInputValue(selectedOption.label);
+  }, [selectedValue, options]);
+
+  // Filter options when input value changes
+  React.useEffect(() => {
+    const selectedOption = options.find(({ value }) => value === selectedValue);
+    if (
+      inputValue &&
+      inputValue.length > 0 &&
+      isOpen &&
+      selectedOption?.label !== inputValue // don't filter list if option is selected
+    ) {
+      const _filteredOptions = options.filter((option) =>
+        option.label.toLowerCase().includes(inputValue?.trim().toLowerCase()),
+      );
+      setFilteredOptions(_filteredOptions);
+      if (!_filteredOptions.find(({ value }) => value === selectedValue)) {
+        setSelectedValue(undefined);
+      }
+    } else {
+      setFilteredOptions(options);
+    }
+  }, [inputValue, options, selectedValue, isOpen]);
+
+  // Update focused value according to filtered list
+  React.useEffect(() => {
+    if (!filteredOptions.find(({ value }) => value === selectedValue)) {
+      setFocusedIndex(options.indexOf(filteredOptions[0]));
+    } else {
+      setFocusedIndex(
+        options.findIndex(({ value }) => value === selectedValue),
+      );
+    }
+  }, [filteredOptions, options, selectedValue]);
+
+  // Update input value and focused value when menu is closed
```

Comment:
I also don't understand what this whole useEffect does.

Reply:
If a value is selected, it closes the menu so I update the `input` value to be the same.

And then I set focused value too, but I might change that

Reply:
Can you just do this on menu item click?

Reply:
there is click and there is enter and there is also external `value` prop

Reply:
I was able to move it into the `onHide` callback of Popover